nuxeo-core/nuxeo-core/src/main/java/org/nuxeo/ecm/core/blob/DocumentBlobProvider.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * (C) Copyright 2015-2017 Nuxeo (http://nuxeo.com/) and others.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Contributors:
+ *     Florent Guillaume
+ */
+package org.nuxeo.ecm.core.blob;
+
+import java.io.IOException;
+import java.io.InputStream;
+
+import org.nuxeo.ecm.core.api.Blob;
+import org.nuxeo.ecm.core.api.DocumentModel;
+import org.nuxeo.ecm.core.model.Document;
+
+/**
+ * Interface for a provider of {@link Blob}s interacting with {@link Document}s.
+ *
+ * @since 9.2
+ */
+public interface DocumentBlobProvider {
+
+    /**
+     * Writes a {@link Blob} to storage and returns information about it.
+     * <p>
+     * Called to store a user-created blob.
+     *
+     * @param blob the blob
+     * @param doc the document to which this blob belongs
+     * @param xpath the xpath of the blob in the document
+     * @return the blob key
+     * @since 9.2
+     */
+    default String writeBlob(Blob blob, Document doc, String xpath) throws IOException {
+        return ((BlobProvider) this).writeBlob(blob);
```

Comment:
Is see your point but in practice this method is only called with a `BlobProvider` that also implements `DocumentBlobProvider`. I didn't want to make `DocumentBlobProvider` extend `BlobProvider` to have clearer callers separation, and this is the only method that mixes both. I've also thought about removing it.
This method is actually here for future extensibility, for potential blob providers that would like to use info about the document or xpath to decide what key to compute. But as it's not really used (or tested) for now, let's remove it.